File: CarHomeWidget/src/main/java/com/anod/car/home/app/MusicAppsActivity.java
package com.anod.car.home.app;

import android.content.AsyncTaskLoader;
import android.content.Context;
import android.content.Intent;
import android.content.Loader;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.os.Bundle;


import com.anod.car.home.BuildConfig;
import com.anod.car.home.CarWidgetApplication;
import com.anod.car.home.R;
import com.anod.car.home.model.AppsList;
import com.anod.car.home.utils.AppLog;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

/**
 * @author alex
 * @date 2014-09-01
 */
public abstract class MusicAppsActivity extends AppsListActivity {
    private static HashSet<String> sExcludePackages;

    static {
        sExcludePackages = new HashSet<String>(2);
        sExcludePackages.add("com.amazon.kindle");
        sExcludePackages.add("com.google.android.apps.magazines");
        sExcludePackages.add("flipboard.app");
        // Samsung crap
        sExcludePackages.add("com.sec.android.app.storycam");
        sExcludePackages.add("com.sec.android.app.mediasync");
        sExcludePackages.add("com.sec.android.mmapp");
        sExcludePackages.add("com.sec.android.automotive.drivelink");
        sExcludePackages.add("com.sec.android.app.mv.player");
        sExcludePackages.add("com.sec.android.app.voicenote");

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    protected int getRowLayoutId() {
        return R.layout.all_apps_row;
    }

    @Override
    protected AppsList getAppList(Context context) {
        return new AppsList(CarWidgetApplication.get(context));
    }


    @Override
    public Loader<ArrayList<AppsList.Entry>> onCreateLoader(int id, Bundle args) {
        return new MediaAppsLoader(this, mAppsList);
    }

    static class MediaAppsLoader extends AsyncTaskLoader<ArrayList<AppsList.Entry>> {

        private final AppsList mAppsList;

        public MediaAppsLoader(Context context, AppsList list) {
            super(context);
            mAppsList = list;
        }

        @Override
        public ArrayList<AppsList.Entry> loadInBackground() {
            final PackageManager packageManager = getContext().getPackageManager();
            List<ResolveInfo> apps = packageManager.queryBroadcastReceivers(new Intent(Intent.ACTION_MEDIA_BUTTON), 96);

            // filter duplicate receivers
            HashMap<String, Boolean> receivers = new HashMap<String, Boolean>();

            for (ResolveInfo appInfo : apps) {
                String pkg = appInfo.activityInfo.packageName;
                // App title
                if (sExcludePackages.contains(pkg) || receivers.containsKey(pkg)) {
                    continue;
                }
                String title = appInfo.activityInfo.applicationInfo.loadLabel(packageManager).toString();
                if (BuildConfig.DEBUG) {
                    AppLog.d(appInfo.activityInfo.packageName + "/" + appInfo.activityInfo.applicationInfo.className);
                }
                receivers.put(pkg, true);
                mAppsList.put(appInfo, title);
            }
            mAppsList.sort();
            return mAppsList.getEntries();
        }
    }
}

File: CarHomeWidget/src/main/java/com/anod/car/home/app/MusicAppsActivity.kt
package com.anod.car.home.app

import com.anod.car.home.BuildConfig
import com.anod.car.home.model.AppsList
import info.anodsplace.framework.AppLog

import android.content.Context
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle

import java.util.HashSet

/**
 * @author alex
 * @date 2014-09-01
 */
abstract class MusicAppsActivity : AppsListActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel.appsList = AppsList()
        viewModel.loaderFactory = { context, callback -> MediaAppsLoader(context, callback) }
    }

    internal class MediaAppsLoader(context: Context, private val callback: AppsListResultCallback) : AsyncTask<Void, Void, List<AppsList.Entry>>() {
        private val packageManager = context.packageManager

        override fun doInBackground(vararg params: Void?): List<AppsList.Entry> {
            val apps = packageManager
                    .queryBroadcastReceivers(Intent(Intent.ACTION_MEDIA_BUTTON), 96)
            // filter duplicate receivers
            val receivers = androidx.collection.SimpleArrayMap<String, Boolean>(apps.size)
            val list = mutableListOf<AppsList.Entry>()
            for (appInfo in apps) {
                val pkg = appInfo.activityInfo.packageName
                // App title
                if (sExcludePackages!!.contains(pkg) || receivers.containsKey(pkg)) {
                    continue
                }
                val title = appInfo.activityInfo.applicationInfo.loadLabel(packageManager)
                        .toString()
                if (BuildConfig.DEBUG) {
                    AppLog.d(appInfo.activityInfo.packageName + "/"
                            + appInfo.activityInfo.applicationInfo.className)
                }
                receivers.put(pkg, true)
                list.add(AppsList.Entry(appInfo, title))
            }
            list.sortBy { it.title }
            return list
        }

        override fun onPostExecute(result: List<AppsList.Entry>?) {
            super.onPostExecute(result)
            callback.onResult(result ?: emptyList())
        }
    }

    companion object {

        private var sExcludePackages: HashSet<String>? = null

        init {
            sExcludePackages = HashSet(2)
            sExcludePackages!!.add("com.amazon.kindle")
            sExcludePackages!!.add("com.google.android.apps.magazines")
            sExcludePackages!!.add("flipboard.app")
            // Samsung crap
            sExcludePackages!!.add("com.sec.android.app.storycam")
            sExcludePackages!!.add("com.sec.android.app.mediasync")
            sExcludePackages!!.add("com.sec.android.mmapp")
            sExcludePackages!!.add("com.sec.android.automotive.drivelink")
            sExcludePackages!!.add("com.sec.android.app.mv.player")
            sExcludePackages!!.add("com.sec.android.app.voicenote")

        }
    }
}
